### PR TITLE
Review branch

### PR DESCRIPTION
**Pull request by nityanandagohain**
Demo

ezgif com-gif-maker

### CODE REVIEW

File: .ruby-version [hunk 1]
```diff
@@ -1 +1 @@
-ruby-2.5.5
+ruby-2.5.1
```
Comment (ankuriitk) [inline .ruby-version]: why downgrade?

File: Gemfile [hunk 2]
```diff
@@ -31,6 +37,8 @@ gem 'bootsnap', '>= 1.4.2', require: false
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
+  gem 'rspec-rails', '~> 4.0.0.beta2'
```
Comment (ankuriitk) [inline Gemfile]: why beta version?

File: .ruby-version (hunk 1 above)
Comment (nityanandagohain) [reply .ruby-version]: was getting problems while deploying to heroku.

File: app/controllers/artists_controller.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        else            
+            puts "already present in database"
+        end
+
+        @artist_details = get_artist_details(@artist.name)
+
+        render "result"
+    end
+
+    def get_artist_details(name)
```
Comment (ankuriitk) [inline app/controllers/artists_controller.rb]: controllers should be thin as  controller code can't be shared. Models  should be  heavy. There is no need for this code to be in controller. Also, any method more than 4-5 lines must have documentation

File: app/controllers/artists_controller.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        # get artist info
+        result = {}
+        url = URI.parse("http://ws.audioscrobbler.com/2.0/?format=json&api_key=#{ENV["LASTFM_KEY"]}&method=artist.getInfo&artist=#{name}")
```
Comment (ankuriitk) [inline app/controllers/artists_controller.rb]: no need to embed constant string. Declare them  as  constant in class and then  use it in code.

File: app/controllers/artists_controller.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        res = Net::HTTP.start(url.host, url.port) { |http|
```
Comment (ankuriitk) [inline app/controllers/artists_controller.rb]: can this call fail?

File: app/controllers/artists_controller.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            http.request(Net::HTTP::Get.new(url.to_s))
+        }
+        data = JSON.parse(res.body)
```
Comment (ankuriitk) [inline app/controllers/artists_controller.rb]: can this fail?

File: Gemfile (hunk 2 above)
Comment (nityanandagohain) [reply Gemfile]: https://github.com/rails/rails/issues/35417

was getting problems so used one of the solutions provided there.

File: app/controllers/artists_controller.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        result["artist"] = data["artist"]
+
+        # get top tracks of the artist
+        url = URI.parse("http://ws.audioscrobbler.com/2.0/?format=json&api_key=#{ENV["LASTFM_KEY"]}&method=artist.gettoptracks&artist=#{name}&limit=10")
```
Comment (ankuriitk) [inline app/controllers/artists_controller.rb]: this method is doing too many things - you can easily split into smaller methods

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
+class UsersController < ApplicationController
+    def index
```
Comment (ankuriitk) [inline app/controllers/users_controller.rb]: why have it if there is no implementation for it?

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  has_secure_password
+
+  validates :email, presence: true, uniqueness: true
+  validates :password, presence: true, length: {minimum: 5}
```
Comment (ankuriitk) [inline app/models/user.rb]: good that you have min length validation

File: app/models/user.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  has_secure_password
+
+  validates :email, presence: true, uniqueness: true
```
Comment (ankuriitk) [inline app/models/user.rb]: no validator for email?

File: db/schema.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 2019_11_18_183028) do
+
+  create_table "artists", force: :cascade do |t|
+    t.string "name"
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.integer "user_id"
```
Comment (ankuriitk) [inline db/schema.rb]: why is this needed?

File: db/schema.rb [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    t.index ["user_id"], name: "index_artists_on_user_id"
+  end
+
+  create_table "users", force: :cascade do |t|
+    t.string "name"
+    t.string "email"
+    t.string "password_digest"
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+  end
```
Comment (ankuriitk) [inline db/schema.rb]: no index?

File: app/controllers/artists_controller.rb (hunk 5 above)
Comment (nityanandagohain) [reply app/controllers/artists_controller.rb]: yes it can, I should have added try catch.

File: spec/models/artist_spec.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe Artist, type: :model do
```
Comment (ankuriitk) [inline spec/models/artist_spec.rb]: test is weak - where are you testing whether code is able to handle the response from last.fm API? You need to have  mocks

File: spec/models/user_spec.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    end
+
+    it 'ensures the password presence' do 
+      user = User.new(email: 'xyz@gmail.com').save
+      expect(user).to eq(false)
+    end
+
+    it 'should save successfully' do
+      user = User.new(email: 'xyz1@xyz.com', password: 'heyypassword').save
+      expect(user).to eq(true)
```
Comment (ankuriitk) [inline spec/models/user_spec.rb]: what if one tries to create two users with same email?

File: db/schema.rb (hunk 12 above)
Comment (nityanandagohain) [reply db/schema.rb]: Added Email as an index for users table.

File: db/schema.rb (hunk 11 above)
Comment (nityanandagohain) [reply db/schema.rb]: Since we are using one to many relationships, a user can search for many artists.

File: app/models/user.rb (hunk 10 above)
Comment (nityanandagohain) [reply app/models/user.rb]: added email validation

File: app/controllers/users_controller.rb (hunk 8 above)
Comment (nityanandagohain) [reply app/controllers/users_controller.rb]: Removed it

File: app/controllers/artists_controller.rb (hunk 3 above)
Comment (nityanandagohain) [reply app/controllers/artists_controller.rb]: Separated the code to and moved them to helper functions

**Comment by nityanandagohain** [comment]
@ankuriitk Hi, thank you for reviewing. I have updated the code can you please have a look at it again?

File: spec/models/user_spec.rb (hunk 14 above)
Comment (nityanandagohain) [reply spec/models/user_spec.rb]: Added teest for it.

File: spec/models/artist_spec.rb (hunk 13 above)
Comment (nityanandagohain) [reply spec/models/artist_spec.rb]: Add it in fastfm_helper_spec.rb

File: .ruby-version (hunk 1 above)
Comment (ankuriitk) [reply .ruby-version]: what problems?

File: app/controllers/artists_controller.rb [hunk 15]
```diff
@@ -0,0 +1,38 @@
+require "net/http"
+require "json"
+
+class ArtistsController < ApplicationController
+    include ArtistsHelper
+    before_action :authorize, only: [:create, :index, :new]
+    Uri = URI.parse("http://ws.audioscrobbler.com/2.0/")
```
Comment (ankuriitk) [inline app/controllers/artists_controller.rb]: why is this here?

File: app/helpers/artists_helper.rb [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        # query params for the api request
+        params = { :format => 'json', :api_key => ENV["LASTFM_KEY"],:method => 'artist.getinfo', :artist => name }
+        uri.query = URI.encode_www_form(params)
+        begin
+            res = Net::HTTP.get_response(uri)
+            if res.is_a?(Net::HTTPSuccess)
+                data = JSON.parse(res.body)
+                return data["artist"]
+            end
+        rescue => exception
```
Comment (ankuriitk) [inline app/helpers/artists_helper.rb]: can you not add retries?

File: app/helpers/artists_helper.rb [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+            puts "Exception Occured while fetching from API"
+            puts exception.backtrace
+        end
+        return nil
+    end
+
+    def getTopTracks(uri, name)
```
Comment (ankuriitk) [inline app/helpers/artists_helper.rb]: each method should have documentation.. also ruby community AFAIK favours underscores over snake case for method/var names

File: db/schema.rb [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    t.index ["user_id"], name: "index_artists_on_user_id"
+  end
+
+  create_table "users", force: :cascade do |t|
+    t.string "name"
+    t.string "email"
+    t.string "password_digest"
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.index ["email"], name: "index_users_on_email"
```
Comment (ankuriitk) [inline db/schema.rb]: is this index right?

File: spec/helpers/fastfm_helper_spec.rb [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+        uri = URI.parse("http://ws.audioscrobbler.com/2.0/")
+        res["artist"] = helpers.getArtistInfo(uri , "Dire Straits")
+        expect(res["artist"].key?("name")).to eq(true)
+        expect(res["artist"].key?("stats")).to eq(true)
+        expect(res["artist"].key?("stats")).to eq(true)
+        expect(res["artist"]["name"]).to eq("Dire Straits")
+      end
+      it "Doesn't get artist info when name is empty" do
+        res = {}
+        uri = URI.parse("http://ws.audioscrobbler.com/2.0/")
```
Comment (ankuriitk) [inline spec/helpers/fastfm_helper_spec.rb]: why repeat the url so many times? Keep things DRY.

File: spec/helpers/fastfm_helper_spec.rb [hunk 20]
```diff
@@ -0,0 +1,34 @@
+require 'rails_helper'
+
+RSpec.describe "FastfmHelper", type: :helper do
+  let(:helpers) { Class.new { extend ArtistsHelper } }
+  describe "FastfmHelper" do
+      it "Get artist info from fastFM api" do
+        res = {}
+        uri = URI.parse("http://ws.audioscrobbler.com/2.0/")
+        res["artist"] = helpers.getArtistInfo(uri , "Dire Straits")
```
Comment (ankuriitk) [inline spec/helpers/fastfm_helper_spec.rb]: where is the mocking happening? Not sure if I understand

File: .ruby-version (hunk 1 above)
Comment (nityanandagohain) [reply .ruby-version]: I got the error of `Your Ruby version is 2.1, but your Gemfile specified 2.5` while deploying to Heroku, but I used Ruby 2.5 for development so downgrading the version in the Gemfile solved the problem.

File: db/schema.rb (hunk 18 above)
Comment (nityanandagohain) [reply db/schema.rb]: Since we are checking if the user exists based on the email, I guess this should do it
